### PR TITLE
feat(jax/array-api): dpa1

### PR DESCRIPTION
**Pull request by njzjz**
## Summary by CodeRabbit

- **New Features**
	- Updated method for converting input to NumPy arrays, enhancing performance and compatibility with array-like structures.
	- Simplified handling of weight, bias, and identity variables for improved compatibility with array backends.
	- Introduced new network classes and enhanced network management functionalities.
	- Added support for the new `array_api_strict` backend in testing.

- **Bug Fixes**
	- Fixed serialization process to ensure accurate conversion of weights and biases.

- **Tests**
	- Added tests to validate the new functionalities and ensure compatibility across various backends, including JAX and Array API Strict.

- **Chores**
	- Continued improvements to project structure and dependencies for better maintainability.

### CODE REVIEW

File: deepmd/dpmodel/utils/region.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 # SPDX-License-Identifier: LGPL-3.0-or-later
+import array_api_compat
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/dpmodel/utils/region.py:2]: Array API doesn't have a type annotation system yet.

File: source/tests/consistent/test_type_embedding.py [hunk 2]
```diff
@@ -72,6 +79,7 @@ def data(self) -> dict:
     dp_class = TypeEmbedNetDP
     pt_class = TypeEmbedNetPT
     jax_class = TypeEmbedNetJAX
+    array_api_strict_class = TypeEmbedNetStrict
     args = type_embedding_args()
     skip_jax = not INSTALLED_JAX
```
Comment (njzjz) [inline source/tests/consistent/test_type_embedding.py:84]:
```suggestion
    skip_jax = not INSTALLED_JAX
    skip_array_api_strict = not INSTALLED_ARRAY_API_STRICT
```

File: deepmd/dpmodel/utils/env_mat.py [hunk 3]
```diff
@@ -44,33 +44,43 @@ def _make_env_mat(
...
+    # index = xp.tile(xp.reshape(nlist, (nf, -1, 1)), (1, 1, 3))
+    # coord_r = xp.take_along_axis(coord,  xp.tile(index, (1, 1, 3)), 1)
+    # note: array api doesn't contain take_along_axis until the next version
+    # reimplement
+    nall = coord.shape[1]
+    index = xp.reshape(nlist, (nf * nloc * nnei,)) + xp.repeat(
+        (xp.arange(nf) * nall), nloc * nnei
+    )
+    coord_ = xp.reshape(coord, (-1, 3))
+    coord_r = xp.take(coord_, index, axis=0)
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/dpmodel/utils/env_mat.py:64]: To my understanding, these positions are masked later

File: source/tests/consistent/test_type_embedding.py [hunk 4]
```diff
@@ -121,6 +130,13 @@ def eval_jax(self, jax_obj: Any) -> Any:
                 raise ValueError("Output is numpy array")
         return [np.array(x) if isinstance(x, jnp.ndarray) else x for x in (out,)]
 
+    def eval_array_api_strict(self, array_api_strict_obj: Any) -> Any:
+        out = array_api_strict_obj()
+        return [
+            np.array(x) if isinstance(x, array_api_strict.ndarray) else x
```
Comment (njzjz) [inline source/tests/consistent/test_type_embedding.py:136]:
```suggestion
            np.asarray(x) if hasattr(x, "__array_namespace__") else x
```

**Review by wanghan-iapcm — COMMENTED** [review]
- Do we change back to `take_along_axis` when it is supported? 
- for modules implemented by jax, you may want to take a look at [flax](https://flax.readthedocs.io/en/latest/)

**Comment by njzjz** [comment]
> * Do we change back to `take_along_axis` when it is supported?

`take_along_axis` has been added to the Array API in https://github.com/data-apis/array-api/pull/816, but I don't expect it to be released quickly. Note that `jax` does support `take_along_axis`, but I hope it can be more general and driven by any Array API library.

However, in https://github.com/deepmodeling/deepmd-kit/pull/4160/commits/d65206f8e6bae8ea7a2b9f6f6f55a991ac4d7674, I add a Python implementation of `take_along_axis`  (per https://github.com/data-apis/array-api/issues/177#issuecomment-2093630595). It also uses `arange` and `take`, which is the same as my previous commits but more general.

> * for modules implemented by jax, you may want to take a look at [flax](https://flax.readthedocs.io/en/latest/)

Is there any advantage to using flax?

File: deepmd/dpmodel/utils/nlist.py [hunk 5]
```diff
@@ -265,36 +277,46 @@ def extend_coord_with_ghosts(
...
+        extend_atype = xp.tile(atype[:, :, xp.newaxis], (1, ns, 1))
+        extend_aidx = xp.tile(aidx[:, :, xp.newaxis], (1, ns, 1))
 
     return (
-        extend_coord.reshape((nf, nall * 3)),
-        extend_atype.reshape((nf, nall)),
-        extend_aidx.reshape((nf, nall)),
+        xp.reshape(extend_coord, (nf, nall * 3)),
+        xp.reshape(extend_atype, (nf, nall)),
+        xp.reshape(extend_aidx, (nf, nall)),
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/dpmodel/utils/nlist.py:321]: `outer` is only available in `linalg` per https://data-apis.org/array-api/latest/extensions/generated/array_api.linalg.outer.html#outer

**Comment by njzjz** [comment]
Here, I make some additional notes for Array API:

- `arr.reshape(shape)` and `arr.astype(dtype)` are not supported; should use `xp.reshape(arr, shape)` and `xp.astype(arr, dtype)`;
- `concatenate` -> `concat`;
- `transpose` -> `permute_dims`;
- `arr[integer_idx]` is not supported; use `xp.take(arr, integer_idx)`;
- `einsum` is not supported.

**Comment by wanghan-iapcm** [comment]
> Is there any advantage to using flax?

jax does not natively support any equivalent to `torch.nn.Module`. `flax` provides something similar, which may be used for better support on parameter initialization and backward propagation.

**Review by wanghan-iapcm — APPROVED** [review]

**Review by iProzd — APPROVED** [review]
